torch/lib/THD/base/channels/DataChannelTCP.cpp

```diff
@@ -196,70 +196,99 @@ DataChannelTCP::~DataChannelTCP()
 
 
 void DataChannelTCP::listen(std::uint16_t port = 0) {
-  SYSCHECK(_socket = ::socket(PF_INET, SOCK_STREAM, 0))
+  struct addrinfo hints, *res = NULL;
+
+  memset(&hints, 0x00, sizeof(hints));
+  hints.ai_flags = AI_PASSIVE;
+  hints.ai_family = AF_UNSPEC; // either IPv4 or IPv6
+  hints.ai_socktype = SOCK_STREAM; // TCP
+
+  // `getaddrinfo` will sort addresses according to RFC 3484 and can be tweeked
+  // by editing `/etc/gai.conf`. so there is no need to manual sorting
+  // or protcol preference.
+  int err = getaddrinfo(NULL, std::to_string(port).data(), &hints, &res);
+  if (err != 0 || !res) {
+    throw std::invalid_argument("cannot find host to listen on: " + std::string(gai_strerror(err)));
+  }
 
-  struct sockaddr_in addr;
-  socklen_t addr_len = sizeof(addr);
+  struct addrinfo *next_addr = res;
+  while (true) {
+    try {
+      SYSCHECK(_socket = ::socket(next_addr->ai_family, next_addr->ai_socktype, next_addr->ai_protocol))
 
-  memset(&addr, 0, addr_len);
-  addr.sin_family = AF_INET;
-  addr.sin_port = htons(port);
-  addr.sin_addr.s_addr = INADDR_ANY;
+      int optval = 1;
+      SYSCHECK(::setsockopt(_socket, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(int)))
+      SYSCHECK(::bind(_socket, next_addr->ai_addr, next_addr->ai_addrlen))
+      SYSCHECK(::listen(_socket, LISTEN_QUEUE_SIZE))
+      break;
+    } catch (const std::system_error& e) {
+      ::close(_socket);
+      next_addr = next_addr->ai_next;
 
-  int optval = 1;
-  SYSCHECK(::setsockopt(_socket, SOL_SOCKET, SO_REUSEADDR, &optval, sizeof(int)))
-  SYSCHECK(::bind(_socket, reinterpret_cast<struct sockaddr*>(&addr), sizeof(addr)))
-  SYSCHECK(::listen(_socket, LISTEN_QUEUE_SIZE))
-  SYSCHECK(::getsockname(_socket, reinterpret_cast<struct sockaddr*>(&addr), &addr_len))
+      // we have tried all addresses but could not establish listening on any of them
+      if (!next_addr) {
+        ::freeaddrinfo(res);
+        throw e;
+      }
+    }
+  }
 
+  // get listen port
+  struct sockaddr_in addr;
+  socklen_t addr_len = sizeof(addr);
+  SYSCHECK(::getsockname(_socket, reinterpret_cast<struct sockaddr*>(&addr), &addr_len))
```

Comment:
It would be simpler if you wrote a `Deleter` class for `addrinfo*` that calls `freeaddrinfo`. You could use a smart pointer and not worry about leaks then.